### ISSUE TITLE
`read_data` function has a mutable default argument

### ISSUE THREAD

**Issue by zmoon** [NCAR/python-tutorial#61]
I think in the description of the function it would be worth adding a sentence about mutable default arguments as [a Python gotcha](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html) that can be troublesome in some cases.

**Comment by kmpaul** [comment]
@zmoon:  Thanks for all your help with this!  We appreciate the contributions, and we hope you will continue to keep us honest!  😄 

Yes.  That is an excellent point.  We should point out that using mutable default arguments is an *anti-pattern* and should be avoided.  Certainly the best practice is to use `None` as the default.  We should add a comment to the text about this, for certain, and maybe we should revisit this material later just to cover more best practices and additional "advanced" Python concepts.  Though, it's fair to argue that best practices should not be considered *advanced*.  😃 

@jukent:  Perhaps after the conclusion of the "Your First..." tutorial series is complete, you could add a comment about "mutable defaults really shouldn't be used."  Best practice is to use something like a "sentinel value" like `None` and then check the value for "noneness" to see if it is unset.

And then after making that addition, we should look over this and create a list of *all* of the commonly used Python knowledge that we haven't covered, such as object-oriented programming and more best practices.

**Comment by kmpaul** [comment]
@zmoon:  If you have a moment, would you mind (in this issue) listing anything that you feel our tutorial is missing?  Anything you think we could improve upon?  Things you hope the tutorial will cover in the future?

**Comment by jukent** [comment]
I will look more into the issues of mutable defaults and add a comment to the tutorial.

**Comment by jukent** [comment]
@zmoon This is a really good point. It seems to be really risky when the function modifies the input argument in some way. Since in our example `types` is never modified, we think this is low risk enough to not explain mutable arguments just yet. Perhaps we should consider changing the function to not set types as a kwarg.

**Comment by zmoon** [comment]
@jukent yeah it's not an issue in this example, as you say. But perhaps it could be mentioned in a "see also", "note" or "warning" box.

I think not having a default, like you suggest, would be fine, since we always pass `types` when we call in the examples. Also the effective default type of `str` could be noted in the docstring. `types={}` doesn't really communicate this, so that could be another reason to remove the default.

Or like @kmpaul said use `None`. Certainly if you make one of these modifications no need to mention the mutable default arguments issue yet.
